Configuration cache fails when using filtered FileCollection

### Current Behavior

The build fails.

### Expected Behavior

The build shouldn't fail.

### Context (optional)

_No response_

### Steps to Reproduce

1. Open the attached project: [FilteredFileCollection.zip](https://github.com/gradle/gradle/files/12589063/FilteredFileCollection.zip)

It has the following build script (note that it uses the `FileCollection#filter` API):
```
abstract class CustomTask : DefaultTask() {

    @get:InputFiles
    abstract val inputFiles: ConfigurableFileCollection

    @get:OutputFile
    abstract val outputFile: RegularFileProperty

    @TaskAction
    fun run() {
        println("Running task $this with inputFiles ${inputFiles.files} and outputFile ${outputFile.get().asFile}")
        outputFile.get().asFile.writeText("Hello")
    }
}

val fooTask by tasks.registering(CustomTask::class) {
    inputFiles.from(project.file("fooInputFile.txt"))
    outputFile.set(project.layout.buildDirectory.file("fooOutputFile.txt"))
}

val barTask by tasks.registering(CustomTask::class) {
    // Set fooTask's outputFile as barTask's inputFiles
    val regularFile: Provider<RegularFile> = fooTask.flatMap { it.outputFile }
    val regularFiles: Provider<List<RegularFile>> = regularFile.map { listOf(it) }
    val fileCollection: FileCollection = project.files(regularFiles)

    inputFiles.from(
        fileCollection
            .filter { it.isFile } // Comment out this line to make the build pass (`./gradlew clean barTask --configuration-cache`)
    )

    outputFile.set(project.layout.buildDirectory.file("barOutputFile.txt"))
}
```
2. Run `./gradlew clean barTask --configuration-cache -s`. The build will fail with:
```
Configuration cache state could not be cached: field `__inputFiles__` of task `:barTask` of type `Build_gradle$CustomTask`: error writing value of type 'org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection'
> Querying the mapped value of flatmap(provider(task 'fooTask', class Build_gradle$CustomTask)) before task ':fooTask' has completed is not supported
Caused by: org.gradle.api.InvalidUserCodeException: Querying the mapped value of flatmap(provider(task 'fooTask', class Build_gradle$CustomTask)) before task ':fooTask' has completed is not supported
        at org.gradle.api.internal.provider.TransformBackedProvider.lambda$beforeRead$0(TransformBackedProvider.java:98)
        at org.gradle.api.internal.provider.ValueSupplier$TaskProducer.visitContentProducerTasks(ValueSupplier.java:140)
        at org.gradle.api.internal.provider.TransformBackedProvider.beforeRead(TransformBackedProvider.java:95)
        at org.gradle.api.internal.provider.TransformBackedProvider.calculateOwnValue(TransformBackedProvider.java:81)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateOwnPresentValue(AbstractMinimalProvider.java:73)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.get(AbstractMinimalProvider.java:93)
        at org.gradle.api.internal.provider.ProviderResolutionStrategy$2.resolve(ProviderResolutionStrategy.java:33)
        at org.gradle.api.internal.file.collections.ProviderBackedFileCollection.visitChildren(ProviderBackedFileCollection.java:64)
        at org.gradle.api.internal.file.CompositeFileCollection$1.visitChildren(CompositeFileCollection.java:103)
        ...
```

### Gradle version

8.3

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

--- a/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheFileCollectionIntegrationTest.groovy
+++ b/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheFileCollectionIntegrationTest.groovy
@@ -212,7 +212,7 @@ class ConfigurationCacheFileCollectionIntegrationTest extends AbstractConfigurat
     }
 
     @Issue('https://github.com/gradle/gradle/issues/26352')
-    def 'ConfigurableFileCollection filter is lazy'() {
+    def 'intricate ConfigurableFileCollection filter chain carries task dependencies'() {
         given:
         def configurationCache = newConfigurationCacheFixture()
 

--- a/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheFileCollectionIntegrationTest.groovy
+++ b/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheFileCollectionIntegrationTest.groovy
@@ -16,6 +16,8 @@
 
 package org.gradle.configurationcache
 
+import spock.lang.Issue
+
 class ConfigurationCacheFileCollectionIntegrationTest extends AbstractConfigurationCacheIntegrationTest {
     def "directory tree is treated as build input when its contents are queried during configuration"() {
         buildFile << """
@@ -207,5 +209,55 @@ class ConfigurationCacheFileCollectionIntegrationTest extends AbstractConfigurat
         "files.files.name"      | "[file1, file2]"
         "files.empty"           | "false"
         "files.contains(file1)" | "true"
+    }
+
+    @Issue('https://github.com/gradle/gradle/issues/26352')
+    def 'ConfigurableFileCollection filter is lazy'() {
+        given:
+        def configurationCache = newConfigurationCacheFixture()
+
+        and:
+        buildFile '''
+            abstract class ConcatTask extends DefaultTask {
+
+                @InputFiles
+                abstract ConfigurableFileCollection getInputFiles()
+
+                @OutputFile
+                abstract RegularFileProperty getOutputFile()
+
+                @TaskAction
+                void run() {
+                    outputFile.get().asFile.text = inputFiles.files.collect { it.text }.join()
+                }
+            }
+
+            def producer = tasks.register('producer', ConcatTask) {
+                inputFiles.from(file("producer-input.txt"))
+                outputFile = layout.buildDirectory.file("producer-output.txt")
+            }
+
+            tasks.register('consumer', ConcatTask) {
+                // Create a complex file collection chain involving `flatMap`, `map` and `filter`
+                Provider<RegularFile> producerOutputFile = producer.flatMap { it.outputFile }
+                Provider<List<RegularFile>> mappedOutputFile = producerOutputFile.map { [it] }
+                ConfigurableFileCollection fileCollection = files(mappedOutputFile)
+
+                inputFiles.from(fileCollection.filter { it.file })
+                outputFile.set(layout.buildDirectory.file("consumer-output.txt"))
+            }
+        '''
+
+        and:
+        file('producer-input.txt').text = '42'
+
+        when:
+        configurationCacheRun 'consumer'
+
+        then:
+        configurationCache.assertStateStored(true)
+
+        and:
+        file('build/consumer-output.txt').text == '42'
     }
 }

--- a/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheIntegrationTest.groovy
+++ b/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheIntegrationTest.groovy
@@ -354,6 +354,7 @@ class ConfigurationCacheIntegrationTest extends AbstractConfigurationCacheIntegr
     }
 
     def "start parameter indicates whether configuration cache was requested"() {
+        given:
         def configurationCache = newConfigurationCacheFixture()
 
         buildFile """

--- a/subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AbstractFileCollection.java
+++ b/subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AbstractFileCollection.java
@@ -71,13 +71,6 @@ public abstract class AbstractFileCollection implements FileCollectionInternal {
         this(DefaultTaskDependencyFactory.withNoAssociatedProject());
     }
 
-    /**
-     * Returns the display name of this file collection. Used in log and error messages.
-     *
-     * @return the display name
-     */
-    public abstract String getDisplayName();
-
     @Override
     public String toString() {
         return getDisplayName();

--- a/subprojects/file-collections/src/main/java/org/gradle/api/internal/file/CompositeFileCollection.java
+++ b/subprojects/file-collections/src/main/java/org/gradle/api/internal/file/CompositeFileCollection.java
@@ -29,7 +29,6 @@ import org.gradle.internal.logging.text.TreeFormatter;
 import java.io.File;
 import java.util.List;
 import java.util.function.Consumer;
-import java.util.function.Supplier;
 
 /**
  * A {@link org.gradle.api.file.FileCollection} that contains the union of zero or more file collections. Maintains file ordering.
@@ -80,32 +79,13 @@ public abstract class CompositeFileCollection extends AbstractFileCollection imp
 
     @Override
     public FileCollectionInternal filter(final Spec<? super File> filterSpec) {
-        return new CompositeFileCollection(taskDependencyFactory, patternSetFactory) {
+        return new FilteredFileCollection(this, filterSpec) {
             @Override
             protected void appendContents(TreeFormatter formatter) {
                 formatter.node("filtered collection");
                 formatter.startChildren();
                 CompositeFileCollection.this.describeContents(formatter);
                 formatter.endChildren();
-            }
-
-            @Override
-            public FileCollectionInternal replace(FileCollectionInternal original, Supplier<FileCollectionInternal> supplier) {
-                FileCollectionInternal newCollection = CompositeFileCollection.this.replace(original, supplier);
-                if (newCollection == CompositeFileCollection.this) {
-                    return this;
-                }
-                return newCollection.filter(filterSpec);
-            }
-
-            @Override
-            protected void visitChildren(Consumer<FileCollectionInternal> visitor) {
-                CompositeFileCollection.this.visitChildren(child -> visitor.accept(child.filter(filterSpec)));
-            }
-
-            @Override
-            public void visitDependencies(TaskDependencyResolveContext context) {
-                CompositeFileCollection.this.visitDependencies(context);
             }
 
             @Override

--- a/subprojects/file-collections/src/main/java/org/gradle/api/internal/file/FileCollectionInternal.java
+++ b/subprojects/file-collections/src/main/java/org/gradle/api/internal/file/FileCollectionInternal.java
@@ -49,6 +49,13 @@ public interface FileCollectionInternal extends FileCollection, TaskDependencyCo
     void visitStructure(FileCollectionStructureVisitor visitor);
 
     /**
+     * Returns the display name of this file collection. Used in log and error messages.
+     *
+     * @return the display name
+     */
+    String getDisplayName();
+
+    /**
      * Appends diagnostic information about the contents of this collection to the given formatter.
      */
     TreeFormatter describeContents(TreeFormatter formatter);

--- a/subprojects/file-collections/src/main/java/org/gradle/api/internal/file/FileTreeInternal.java
+++ b/subprojects/file-collections/src/main/java/org/gradle/api/internal/file/FileTreeInternal.java
@@ -24,8 +24,6 @@ import java.util.function.Consumer;
 public interface FileTreeInternal extends FileTree, FileCollectionInternal {
     String DEFAULT_TREE_DISPLAY_NAME = "file tree";
 
-    String getDisplayName();
-
     void visitContentsAsFileTrees(Consumer<FileTreeInternal> visitor);
 
     @Override

--- a/subprojects/file-collections/src/main/java/org/gradle/api/internal/file/FilteredFileCollection.java
+++ b/subprojects/file-collections/src/main/java/org/gradle/api/internal/file/FilteredFileCollection.java
@@ -43,7 +43,7 @@ public class FilteredFileCollection extends AbstractFileCollection {
         if (newCollection == collection) {
             return this;
         }
-        return new FilteredFileCollection(newCollection, filterSpec);
+        return newCollection.filter(filterSpec);
     }
 
     public FileCollectionInternal getCollection() {
@@ -56,7 +56,7 @@ public class FilteredFileCollection extends AbstractFileCollection {
 
     @Override
     public String getDisplayName() {
-        return "file collection";
+        return collection.getDisplayName();
     }
 
     @Override

--- a/subprojects/file-collections/src/test/groovy/org/gradle/api/internal/file/CompositeFileCollectionTest.groovy
+++ b/subprojects/file-collections/src/test/groovy/org/gradle/api/internal/file/CompositeFileCollectionTest.groovy
@@ -220,8 +220,6 @@ class CompositeFileCollectionTest extends Specification {
     def "can filter the elements of collection"() {
         def source1 = Mock(FileCollectionInternal)
         def source2 = Mock(FileCollectionInternal)
-        def filtered1 = Mock(FileCollectionInternal)
-        def filtered2 = Mock(FileCollectionInternal)
         def spec = Stub(Spec)
         def collection = new TestCompositeFileCollection(source1, source2)
 
@@ -231,20 +229,13 @@ class CompositeFileCollectionTest extends Specification {
         then:
         0 * _
 
-        when:
-        def result = filtered.getSourceCollections()
-
-        then:
-        result == [filtered1, filtered2]
-        1 * source1.filter(spec) >> filtered1
-        1 * source2.filter(spec) >> filtered2
-        0 * _
+        and:
+        filtered instanceof FilteredFileCollection
     }
 
     def "can replace backing collection of filtered collection"() {
         def source1 = Mock(FileCollectionInternal)
         def source2 = Mock(FileCollectionInternal)
-        def filtered2 = Mock(FileCollectionInternal)
         def other = Mock(FileCollectionInternal)
         def spec = Stub(Spec)
         def collection = new TestCompositeFileCollection(source1, source2)
@@ -259,14 +250,6 @@ class CompositeFileCollectionTest extends Specification {
         then:
         replaced.is(filtered)
         replaced2 != filtered
-        0 * _
-
-        when:
-        def result = replaced2.getSourceCollections()
-
-        then:
-        result == [filtered2]
-        1 * source2.filter(spec) >> filtered2
         0 * _
     }
 

--- a/subprojects/file-collections/src/test/groovy/org/gradle/api/internal/file/collections/ProviderBackedFileCollectionTest.groovy
+++ b/subprojects/file-collections/src/test/groovy/org/gradle/api/internal/file/collections/ProviderBackedFileCollectionTest.groovy
@@ -51,7 +51,7 @@ class ProviderBackedFileCollectionTest extends Specification {
         result == [task] as Set
     }
 
-    def "resolves task dependencies for provide with unknown producer and file value"() {
+    def "resolves task dependencies for provider with unknown producer and file value"() {
         when:
         def dependencies = fileCollection.buildDependencies
 
@@ -67,7 +67,7 @@ class ProviderBackedFileCollectionTest extends Specification {
         result.empty
     }
 
-    def "resolves task dependencies for provide with unknown producer and buildable value"() {
+    def "resolves task dependencies for provider with unknown producer and buildable value"() {
         def task = Stub(Task)
         def value = Mock(Buildable)
 

